File: backend/src/main/java/com/halisaha/reservation/entity/Reservation.java
package com.halisaha.reservation.entity;

import com.halisaha.field.entity.Field;
import com.halisaha.payment.PaymentMethod;
import com.halisaha.payment.ReservationPaymentStatus;
import com.halisaha.reservation.ReservationStatus;
import com.halisaha.user.entity.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.ZonedDateTime;

@Entity
@Table(name = "reservations")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "confirmation_code", unique = true, nullable = false, length = 20)
    private String confirmationCode;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "field_id", nullable = false)
    private Field field;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @Column(name = "game_type", nullable = false, length = 50)
    private String gameType;

    @Column(name = "guest_name", length = 100)
    private String guestName;

    @Column(name = "guest_phone", length = 20)
    private String guestPhone;

    @Column(name = "guest_email", length = 255)
    private String guestEmail;

    @Column(name = "manage_token_hash", length = 64)
    private String manageTokenHash;

    @Column(name = "start_time", nullable = false)
    private ZonedDateTime startTime;

    @Column(name = "end_time", nullable = false)
    private ZonedDateTime endTime;

    @Column(name = "duration_minutes", nullable = false)
    private Integer durationMinutes;

    @Column(name = "total_price", nullable = false, precision = 10, scale = 2)
    private BigDecimal totalPrice;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private ReservationStatus status = ReservationStatus.CONFIRMED;

    @Column(name = "privacy_accepted", nullable = false)
    @Builder.Default
    private Boolean privacyAccepted = false;

    @Column(name = "privacy_accepted_at")
    private ZonedDateTime privacyAcceptedAt;

    @Column(name = "notification_consent")
    @Builder.Default
    private Boolean notificationConsent = false;

    @Column(name = "notification_consent_at")
    private ZonedDateTime notificationConsentAt;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_status", length = 20)
    @Builder.Default
    private ReservationPaymentStatus paymentStatus = ReservationPaymentStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", length = 20)
    private PaymentMethod paymentMethod;

    @Column(name = "stripe_payment_intent_id", length = 255)
    private String stripePaymentIntentId;

    @Column(name = "paid_at")
    private ZonedDateTime paidAt;

    @Column(name = "refunded_at")
    private ZonedDateTime refundedAt;

    @Column(name = "cancelled_at")
    private ZonedDateTime cancelledAt;

    @Column(name = "cancelled_by", length = 20)
    private String cancelledBy;

    @Column(name = "coupon_code", length = 50)
    private String couponCode;

    @Column(name = "discount_amount", precision = 10, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private ZonedDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private ZonedDateTime updatedAt;
}

File: frontend/src/types/reservation.ts
export type ReservationStatus = 'CONFIRMED' | 'CANCELLED' | 'MODIFIED' | 'COMPLETED' | 'NO_SHOW'
export type PaymentStatus = 'PENDING' | 'PAID' | 'FAILED' | 'REFUNDED' | 'ON_SITE'
export type PaymentMethod = 'CARD' | 'APPLE_PAY' | 'GOOGLE_PAY' | 'ON_SITE'

export interface Reservation {
  id: number
  confirmationCode: string
  fieldId: number
  fieldName: string
  fieldType: string
  customerName: string
  customerPhone: string
  customerEmail: string
  manageToken?: string | null
  startTime: string
  endTime: string
  durationMinutes: number
  totalPrice: number
  status: ReservationStatus
  paymentStatus: PaymentStatus
  paymentMethod?: PaymentMethod
  createdAt: string
  equipmentRentals?: EquipmentRental[]
}

export interface CreateReservationRequest {
  fieldId: number
  startTime: string
  durationMinutes: number
  guestName?: string
  guestPhone?: string
  guestEmail?: string
  privacyAccepted: boolean
  notificationConsent?: boolean
  equipmentRentals?: { equipmentId: number; quantity: number; size?: string }[]
  sessionId?: string
}

export interface ModifyReservationRequest {
  startTime: string
  durationMinutes: number
}

export interface EquipmentRental {
  equipmentName: string
  quantity: number
  size?: string
  price: number
}

export interface SlotHoldRequest {
  fieldId: number
  startTime: string
  durationMinutes: number
  sessionId: string
}
